### ISSUE TITLE
Faster implementation of monitor than standard library

### ISSUE THREAD

**Issue by ganmacs** [fluent/fluentd#2645]
I'll try https://github.com/nurse/ext_monitor

cc. @nurse

**Comment by ganmacs** [comment]
oh… I made a mistake… transfered to fluent/fluentd.

**Comment by repeatedly** [comment]
Here is my benchmark result of https://github.com/fluent/fluentd/pull/2561
`Monitor#synchronize` is still slower...

```
- ruby 2.6.3

Warming up --------------------------------------
             monitor   184.967k i/100ms
               mutex   289.882k i/100ms
Calculating -------------------------------------
             monitor      3.287M (± 2.2%) i/s -     16.462M in   5.011436s
               mutex      8.005M (± 1.6%) i/s -     40.294M in   5.034949s

- ruby 2.6.5
Warming up --------------------------------------
             monitor    56.806k i/100ms
               mutex   280.742k i/100ms
Calculating -------------------------------------
             monitor    678.492k (± 4.7%) i/s -      3.408M in   5.035571s
               mutex      8.054M (± 2.4%) i/s -     40.427M in   5.022535s

- ruby 2.6.5 with ext_monitor
Warming up --------------------------------------
             monitor    62.786k i/100ms
               mutex   292.756k i/100ms
Calculating -------------------------------------
             monitor    754.165k (± 2.2%) i/s -      3.830M in   5.080977s
               mutex      8.031M (± 1.4%) i/s -     40.400M in   5.031691s
```

**Comment by ganmacs** [comment]
For my understanding, this gem is a faster version of the exising `Monitor` class, not `Mutex` class.
So we should compare Monitor library ruby provides and this gem.

```rb
require 'benchmark/ips'
# require 'ext_monitor'

mo = Monitor.new
mu = Mutex.new

p "Ruby #{RUBY_VERSION}"

Benchmark.ips do |x|
  x.report("monitor") do
    mo.synchronize {}
  end

  x.report("mutex") do
    mu.synchronize {}
  end
end
```

with ext_monitor
```
"Ruby 2.5.5"
Warming up --------------------------------------
             monitor   266.565k i/100ms
               mutex   351.016k i/100ms
Calculating -------------------------------------
             monitor      4.937M (± 4.7%) i/s -     24.791M in   5.036034s
               mutex      7.851M (± 4.7%) i/s -     39.314M in   5.018843s
```

without ext_monitor (commented out)
```
"Ruby 2.5.5"
Warming up --------------------------------------
             monitor   228.634k i/100ms
               mutex   347.341k i/100ms
Calculating -------------------------------------
             monitor      3.797M (± 2.1%) i/s -     18.977M in   5.000240s
               mutex      8.139M (± 1.8%) i/s -     40.986M in   5.037270s
```

**Comment by repeatedly** [comment]
Yes, but the problem is ruby 2.6.4 and later has slower Monitor.
See my benchmark result. ruby 2.6.5 with ext_monitor is still very slower than ruby 2.6.3 without ext_monitor.

**Comment by ganmacs** [comment]
ah, my bad. You already compared them...
The benchmark running on my local MacBook results in a similar one to yours (around 838.033**k** even if using ext_monitor).

**Comment by nurse** [comment]
Thank you for testing. I found a bug and fixed and released v0.1.1: https://rubygems.org/gems/ext_monitor/versions/0.1.1

**Comment by ganmacs** [comment]
Great 👏 

with ext_monitor (2x faster)

```
"Ruby 2.6.3"
Warming up --------------------------------------
             monitor   327.029k i/100ms
               mutex   362.710k i/100ms
Calculating -------------------------------------
             monitor      7.507M (± 2.4%) i/s -     37.608M in   5.012856s
               mutex     10.517M (± 2.5%) i/s -     52.593M in   5.004291s
```

without ext_monitor

```
"Ruby 2.6.3"
Warming up --------------------------------------
             monitor   230.079k i/100ms
               mutex   364.688k i/100ms
Calculating -------------------------------------
             monitor      3.781M (± 4.7%) i/s -     18.866M in   5.000347s
               mutex      9.479M (± 5.1%) i/s -     47.409M in   5.014644s
```


with ext_monitor(10x faster)
```
"Ruby 2.6.5"
Warming up --------------------------------------
             monitor   333.879k i/100ms
               mutex   377.950k i/100ms
Calculating -------------------------------------
             monitor      8.036M (± 3.1%) i/s -     40.399M in   5.033352s
               mutex     10.204M (± 3.9%) i/s -     51.023M in   5.008424s
```

without ext_monitor
```
"Ruby 2.6.5"
Warming up --------------------------------------
             monitor    73.338k i/100ms
               mutex   385.539k i/100ms
Calculating -------------------------------------
             monitor    892.246k (± 5.1%) i/s -      4.474M in   5.028408s
               mutex     10.528M (± 2.6%) i/s -     52.819M in   5.020783s
```
